open-log/rfcs/0001-storage.md

```diff
@@ -0,0 +1,154 @@
+# RFC 0001: Open-Log Storage
+
+**Status**: Draft
+
+**Authors**:
+- Jason Gustafson <12502538+hachikuji@users.noreply.github.com>
+
+## Summary
+
+This RFC defines the storage model and core API for open-log. Keys are stored directly in the LSM with a sequence number suffix, enabling per-key log streams with global ordering. The API provides append and scan operations mirroring SlateDB's interface.
+
+## Motivation
+
+Open-Log is a key-oriented log system inspired by Kafka, but with a simpler model: keys are user-defined and every key is logically its own independent log. There is no concept of partitions or repartitioning—users simply write to new keys when their access patterns change.
+
+Logs are stored in SlateDB's LSM tree. Writes are appended to the WAL and memtable, then flushed to sorted string tables (SSTs). LSM compaction naturally organizes data for log locality, grouping entries by key prefix over time. This provides efficient sequential reads for individual logs without requiring explicit partitioning infrastructure.
+
+## Goals
+
+- Define a write API for appending entries to a key's log
+- Define a scan API for reading entries from a key's log
+
+## Non-Goals (left for future RFCs)
+
+- Active polling or push-based consumption
+- Retention policies
+- Checkpoints and cloning
+- Key-range scans
+
+## Design
+
+### Key Encoding
+
+SlateDB keys are a composite of the user key and a `u64` sequence number. A leading byte discriminator distinguishes entry types.
+
+```
+Log Entry:
+  SlateDB Key:   | 0x01 | key (bytes) | sequence (u64) |
+  SlateDB Value: | record value (bytes) |
```

Comment:
do we want to consider headers from day 0? I feel like this caused a lot of pain in Kafka but I don't have the context as to why. I'm not totally sure I see why headers should be something special.

Reply:
I added headers in the list of alternatives. I think it's a good example that demonstrates the kind of simplicity tradeoff that we expect in opendata structures.